cmd/soroban-rpc/internal/db/migration.go

```diff
@@ -0,0 +1,183 @@
+package db
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/config"
+)
+
+type LedgerSeqRange struct {
+	firstLedgerSeq uint32
+	lastLedgerSeq  uint32
+}
+
+func (mlr *LedgerSeqRange) IsLedgerIncluded(ledgerSeq uint32) bool {
+	if mlr == nil {
+		return false
+	}
+	return ledgerSeq >= mlr.firstLedgerSeq && ledgerSeq <= mlr.lastLedgerSeq
+}
+
+func (mlr *LedgerSeqRange) Merge(other *LedgerSeqRange) *LedgerSeqRange {
+	if mlr == nil {
+		return other
+	}
+	if other == nil {
+		return mlr
+	}
+	return &LedgerSeqRange{
+		firstLedgerSeq: min(mlr.firstLedgerSeq, other.firstLedgerSeq),
+		lastLedgerSeq:  max(mlr.lastLedgerSeq, other.lastLedgerSeq),
+	}
+}
+
+type MigrationApplier interface {
+	Apply(ctx context.Context, meta xdr.LedgerCloseMeta) error
+	// ApplicableRange returns the closed ledger sequence interval,
+	// a null result indicates the empty range
+	ApplicableRange() *LedgerSeqRange
+}
+
+type migrationApplierFactory interface {
+	New(db *DB, latestLedger uint32) (MigrationApplier, error)
+}
+
+type migrationApplierFactoryF func(db *DB, latestLedger uint32) (MigrationApplier, error)
+
+func (m migrationApplierFactoryF) New(db *DB, latestLedger uint32) (MigrationApplier, error) {
+	return m(db, latestLedger)
+}
+
+type Migration interface {
+	MigrationApplier
+	Commit(ctx context.Context) error
+	Rollback(ctx context.Context) error
+}
+
+type multiMigration []Migration
+
+func (mm multiMigration) ApplicableRange() *LedgerSeqRange {
+	var result *LedgerSeqRange
+	for _, m := range mm {
+		result = m.ApplicableRange().Merge(result)
+	}
+	return result
+}
+
+func (mm multiMigration) Apply(ctx context.Context, meta xdr.LedgerCloseMeta) error {
+	var err error
+	for _, m := range mm {
+		if localErr := m.Apply(ctx, meta); localErr != nil {
+			err = errors.Join(err, localErr)
+		}
+	}
+	return err
+}
+
+func (mm multiMigration) Commit(ctx context.Context) error {
+	var err error
+	for _, m := range mm {
+		if localErr := m.Commit(ctx); localErr != nil {
+			err = errors.Join(err, localErr)
+		}
+	}
+	return err
+}
+
+func (mm multiMigration) Rollback(ctx context.Context) error {
+	var err error
+	for _, m := range mm {
+		if localErr := m.Rollback(ctx); localErr != nil {
+			err = errors.Join(err, localErr)
+		}
+	}
+	return err
+}
+
+// guardedMigration is a db data migration whose application is guarded by a boolean in the meta table
+// (after the migration is applied the boolean is set to true, so that the migration is not applied again)
+type guardedMigration struct {
```

Comment:
Could we guard against duplicate migrations by comparing the `ledger_sequence` range in the transactions table against ledger meta rather than this method? Because wouldn't this prevent the migration from happening on subsequent releases? So if someone upgrades to 21.3.0 then 21.3.1 they wouldn't update their transactions table?

Reply:
> Could we guard against duplicate migrations by comparing the ledger_sequence range in the transactions table against ledger meta rather than this method?

The length of meta doesn't need to match the length of the transactions due to the configurable window size. What's more, it can have changed several times.

> Because wouldn't this prevent the migration from happening on subsequent releases? So if someone upgrades to 21.3.0 then 21.3.1 they wouldn't update their transactions table?

I am not sure I understand. Upgrades should only happen once.

Reply:
Lemme redirect my question to something simpler so that I can understand the flow here: when does the "the migration happen" meta key get set back to `false`? Or do we need new meta keys every time we need a migration?

Reply:
There is one meta key per migration

Reply:
We could maybe use a dedicated table for the data migration keys (instead of the meta table). But I doubt that's a considerable improvement (famous last words)